nexus/src/app/saga.rs

```diff
@@ -135,22 +185,143 @@ impl super::Nexus {
+    /// which can be cancelled.  **This function should not be used in a
+    /// `tokio::select!` with a `timeout` or the like.**
+    ///
+    /// Say you _do_ want to kick off a saga and wait only a little while before
+    /// it completes.  In that case, you can use the lower-level interface to
+    /// first create the saga (a process which still should not be cancelled,
+    /// but would generally be quick) and then wait for it to finish.  The
+    /// waiting part is cancellable.
+    ///
+    /// Note that none of this affects _crash safety_.  In terms of a crash: the
+    /// crash will either happen before the saga has been created (in which
+    /// case it's as though we didn't even call this function) or after (in
+    /// which case the saga will run to completion).
+    pub(crate) async fn saga_execute<N: NexusSaga>(
         &self,
-        runnable_saga: RunnableSaga,
+        params: N::Params,
     ) -> Result<SagaResultOk, Error> {
-        let log = &self.log;
+        // Construct the DAG specific to this saga.
+        let dag = create_saga_dag::<N>(params)?;
+        let runnable_saga = self.saga_prepare(dag).await?;
+        let running_saga = runnable_saga.start().await?;
+        let stopped_saga = running_saga.wait_until_stopped().await;
+        stopped_saga.into_omicron_result()
+    }
+}
+
+/// Encapsulates a saga to be run before we actually start running it
+///
+/// At this point, we've built the DAG, loaded it into the SEC, etc. but haven't
+/// started it running.  This is a useful point to inject errors, inspect the
+/// DAG, etc.
+pub(crate) struct RunnableSaga {
+    id: SagaId,
+    saga_completion_future: BoxFuture<'static, SagaResult>,
+    log: slog::Logger,
+    sec_client: Arc<steno::SecClient>,
+}
+
+impl RunnableSaga {
+    #[cfg(test)]
+    pub(crate) fn id(&self) -> SagaId {
+        self.id
+    }
+
+    /// Start this saga running.
+    pub(crate) async fn start(self) -> Result<RunningSaga, Error> {
+        info!(self.log, "starting saga");
         self.sec_client
-            .saga_start(runnable_saga.id)
+            .saga_start(self.id)
             .await
             .context("starting saga")
             .map_err(|error| Error::internal_error(&format!("{:#}", error)))?;
 
-        let result = runnable_saga.fut.await;
-        result.kind.map_err(|saga_error| {
+        Ok(RunningSaga {
+            id: self.id,
+            saga_completion_future: self.saga_completion_future,
+            log: self.log,
+        })
+    }
+}
+
+/// Describes a saga that's been started running
+pub(crate) struct RunningSaga {
+    id: SagaId,
+    saga_completion_future: BoxFuture<'static, SagaResult>,
+    log: slog::Logger,
+}
+
+impl RunningSaga {
+    /// Waits until this saga stops executing
+    ///
+    /// Normally, the saga will have finished successfully or failed and unwound
+    /// completely.  If unwinding fails, it will be _stuck_ instead.
```

Comment:
To be clear, will this "stuck" state show up as `Stopped`, or no?

(Based on the log below, I'm assuming "stuck sagas" still appear "stopped", but wanted to clarify this)

Reply:
It depends what you mean.  You will still get a `StoppedSaga` object and the saga will indeed not be running any more (ever).  If you fetch the state of the saga (e.g., using `Nexus::saga_get()`), its state should be [`SagaState::Stuck`](https://github.com/oxidecomputer/omicron/blob/8accfd3767679d3908bc12e0209b04b9ea6c5b8a/nexus/types/src/internal_api/views.rs#L46-L67).  There is no "stopped" state in that enum -- we break out succeeded vs. failed vs. stuck.

Reply:
That makes sense, I just wanted to clarify the convention of this specific function, named `wait_until_stopped`. 

The comment that:
- "Normally, the saga will have finished successfully or failed and unwind..."
- "... (but) if unwinding fails, it will be stuck instead"

Perhaps it's the use of the word "instead", but this made me wonder if "getting stuck" would cause this function to return. From our conversation, I can see the answer is "yes, it will still return a `StoppedSaga`, but that wasn't immediately clear to me from the comment.

Reply:
Sure, I'm happy to update this.  Is it the comment or the name of the function that's confusing?

Basically, once the saga has been started, Steno will run it until it comes to rest, at which point one of a few things will have happened:

* The saga completed successfully.  That means that all actions ran and produced no errors.  This is `SagaState::Succeeded`.
* The saga encountered an error, then unwound successfully.  That means some of the actions ran, one or more actions produced an error, then all of the successful actions' undo actions also ran, and there were _no_ errors from the undo actions.  This is `SagaState::Failed`.
* The saga encountered an error, then started unwinding, then encountered another error from one of the undo actions.  This is `SagaState::Stuck` because it is (forever) stuck neither having fully run all the actions nor having fully unwound them.

In terms of the _name_ of the function: I went with "stopped" as a sort of ambiguous word that could cover all three of these cases, figuring the comment would explain it more precisely.  I also considered "done" (as in `wait_until_done()`) and "finished" but those kind of suggest success to me (so I'm not sure they cover the "failed" case that well) and they certainly seem surprising for the "stuck" state.  There's "paused", but that suggests it might start up again, which is not the case.  I'm open to other words here.

In terms of the comment, how about this:

> Waits until the saga stops executing
>
> This function waits until the saga stops executing because one of the following three things happens:
>
> 1. The saga completes successfully ([`SagaState::Success`]).
> 2. The saga fails and unwinding completes without errors ([`SagaState::Failed`]).
> 3. The saga fails and then an error is encountered during unwinding ([`SagaState::Stuck`]).
>
> Steno continues running the saga (and this function continues waiting) until one of those three things happens.  Once any of those things happens, the saga is no longer running and this function returns a `StoppedSaga` that you can use to inspect more precisely what happened.
```

Reply:
I think the term "stopped" makes sense here, no grip with the function name! I do think that comment would clarify the question I had very clearly, since it calls out very explicitly what "stopped" means.

Reply:
Updating in #5974